maven_iuvs/echelle.py

```diff
@@ -0,0 +1,886 @@
+
+    # Lastly, filter by date/time
+    if date is not None:
+
+        # To get observations for a range of dates:
+        if type(date) is list:
+            if type(date[0]) == datetime.date: # If no time information was entered, be liberal and assume start of first day and end of last
+                date[0] = datetime.datetime(date[0].year, date[0].month, date[0].day, 0, 0, 0)
+
+            if type(date[1]) == datetime.date:
+                date[1] = datetime.datetime(date[1].year, date[1].month, date[1].day, 23, 59, 59)
+            elif date[1] == -1: # Use this to just go until the present time/date.
+                date[1] = datetime.datetime.utcnow()
+            
+            print(f"Returning observations between {date[0]} and {date[1]}")
+
+            selected_lights = [entry for entry in selected_lights if date[0] <= entry['datetime'] <= date[1]]
+
+        # To get observations at a specific day or specific day/time:
+        elif type(date) is not list:  
+
+            if type(date) == datetime.date: # If no time information was entered, be liberal and assume start of first day and end of last
+                date0 = datetime.datetime(date.year, date.month, date.day, 0, 0, 0)
+                date1 = datetime.datetime(date.year, date.month, date.day, 23, 59, 59)
+
+                selected_lights = [entry for entry in selected_lights if date0 <= entry['datetime'] <= date1]
+
+            else: # if a full datetime.datetime object is entered, look for that exact entry.
+                selected_lights = [entry for entry in selected_lights if entry['datetime'] == date]
+
+        else:
+            raise TypeError(f"Date entered is of type {type(date)}")
+
+    return selected_lights
+
+# Relating to dark vs. light observations -----------------------------
+
+
+def coadd_lights(light_fits, dark_fits):
+    """
+    Co-add all light frames within light_fits, including subtraction
+    of dark frames in dark_fits.
+
+    Parameters
+    ----------
+    light_fits : IUVSFits object or HDUlist
+                 fits object representing the light observations.
+    dark_fits : IUVSFits object or HDUlist
+                 fits object representing the associated darks.
+
+    Returns
+    ----------
+    coadded_lights : array
+                     Essentially the mean-frame of the detector
+    nan_light_inds : array
+                     Indices of light frames in which there are nan values.
+    bad_light_inds : array
+                     Indices of light frames in which there are bad values (not nans)
+    nan_dark_inds : array
+                    Indices of dark frames in which there are nans.
+    total_frames : int
+                   total number of good frames which were used to create the 
+                   co-added image. Returned so the value can be added to 
+                   quicklooks.
+    """
+
+    # Retrieve dark frames for subtraction
+    first_dark, second_dark = get_dark_frames(dark_fits)
```

Comment:
I think assuming that there is always a first and second dark will cause problems when processing the very earliest mission data, but we can deal with that later.

Reply:
Right now it throws an error if there's only 1, returns the 2 if there are 2, and returns 2 with the second being an average of 2:end if there's > 2. I can change the unpacking to a tuple at a later point though

Reply:
Started on this with bcddfdf576c7739e451c3c8ae3a396c4dd0193ae